hollow-test/src/test/java/com/netflix/hollow/test/consumer/TestHollowConsumerTest.java

```diff
@@ -98,4 +102,103 @@ public void testAddSnapshot_afterUpdate() throws Exception {
         consumer.triggerRefresh();
         assertEquals(version2, consumer.getCurrentVersionId());
     }
+
+    @Test
+    public void testAddDelta_versionTransition() throws Exception {
+        long snapshotVersion = 1L;
+        TestHollowConsumer consumer = new TestHollowConsumer.Builder()
+                .withBlobRetriever(new TestBlobRetriever())
+                .build();
+
+        HollowWriteStateEngine snapshotWriteStateEngine = new HollowWriteStateEngineBuilder().build();
+        consumer.addSnapshot(snapshotVersion, snapshotWriteStateEngine);
+        assertEquals("Should be no version", VERSION_NONE, consumer.getCurrentVersionId());
+        consumer.triggerRefreshTo(snapshotVersion);
+        assertEquals("Should be at snapshot version", snapshotVersion, consumer.getCurrentVersionId());
+
+        long toVersion = 2L;
+        HollowWriteStateEngine deltaWriteStateEngine = new HollowWriteStateEngineBuilder().build();
+        consumer.addDelta(snapshotVersion, toVersion, snapshotWriteStateEngine, deltaWriteStateEngine);
+        consumer.triggerRefreshTo(toVersion);
+        assertEquals("Should be at delta version", toVersion, consumer.getCurrentVersionId());
+    }
+
+    @Test
+    public void testAddDelta_dataTransition() throws Exception {
+
+        TestHollowConsumer consumer = new TestHollowConsumer.Builder()
+                .withBlobRetriever(new TestBlobRetriever())
+                .build();
+
+        // snapshot transition to version 1
+        long v1 = 1L;
+        HollowWriteStateEngine snapshotWriteStateEngine = new HollowWriteStateEngineBuilder(Collections.singletonList(Movie.class))
+                .add(new Movie(1, "foo"))
+                .build();
+        consumer.addSnapshot(v1, snapshotWriteStateEngine);
+        consumer.triggerRefreshTo(v1);
+        HollowDataAccess data = consumer.getAPI().getDataAccess();
+        assertEquals("Should have Movie and String", new HashSet<>(Arrays.asList("Movie", "String")), data.getAllTypes());
+        customAssertEquals(data, 0, new Movie(1, "foo"));
+        assertEquals("foo", new GenericHollowObject(data, "String", 0).getString("value"));
+
+        // delta transition to version 2
+        long v2 = 2L;
+        HollowWriteStateEngine deltaWriteStateEngine = new HollowWriteStateEngineBuilder(Collections.singletonList(Movie.class))
+                .restoreFrom(consumer.getStateEngine())
+                .add(new Movie(2, "bar"))
+                .build();
+        consumer.addDelta(consumer.getCurrentVersionId(), v2, snapshotWriteStateEngine, deltaWriteStateEngine);
+        consumer.triggerRefreshTo(v2);
+        data = consumer.getAPI().getDataAccess();
+        customAssertEquals(data, 0, new Movie(1, "foo"));   // ghost record
```

Comment:
I don't understand why this is a ghost record - we restored from v1 and added `Movie(2, "bar")` without removing `Movie(1,  "foo")`, right? I'm probably misunderstanding what `deltaWriteStateEngine` is and how `restoreFrom` works...

Reply:
`restoreFrom` is the same code path that a producer calls when restoring prior read state to continue the delta chain. 
`deltaWriteStateEngine` is similar to the workable write state for the producer cycle after restoring from prior read state.
Any records that aren't re-added to the write state will be considered removed.

Reply:
I see, thanks! It might be possible to simplify the API to something like this:
```
HollowStateEngine engine1 = new HollowWriteStateEngineBuilder(...);
HollowStateEngine engine2 = new HollowWriteStateEngineBuilder(...);
long v1 = 1L;
long v2 = 2L;
consumer.addSnapshot(v1, engine1);
consumer.addDelta(v1, v2, engine2);
```
Where you _must_ `addSnapshot(v1)` before `addDelta(v1, v2)`. You could also choose to `consumer.addSnapshot(v2, engine2)` after, if that's not too complex to implement.
We can handle the `withRestore` and `previousStateEngine` bits inside the addDelta call ourselves since we already have the state engine for the "from" version (which could be a snapshot state engine _or_ a delta state engine) and throw a helpful exception if the user added state engines in the wrong order.

Reply:
good suggestions, implemented